### PR TITLE
Multiple error array

### PR DESCRIPTION
**Pull request by graceolivia**
Added new method `InputValidator.validateEachStitchInWholePattern` which is called to find both stitch count errors and invalid stitch type errors.

Removed multiple-stitch error checking from `NestedArrayBuilder` and adjusted `NestedArrayBuilder` tests accordingly.

Added new func `isStitchAndStitchCountValid` in StitchLibrary.swift which returns a result with an `InputError` of type `invalidStitchNumber` or `invalidStitch` which is called in `InputValidator`.

### CODE REVIEW

File: Sources/SwiftCliCore/StitchLibrary.swift [hunk 1]
```diff
@@ -50,3 +48,33 @@ func isStitchValid(stitch: String) -> Bool {
...
+
+func isStitchAndStitchCountValid(stitch: String, rowNumber: Int, stitchIndex: Int) -> Result<String, InputError> {
+    if !isStitchValid(stitch: stitch) {
+        return .failure(InputError.invalidStitch(invalidStitch: stitch, rowLocation: rowNumber, stitchIndexInRow: stitchIndex))
+    }
+    if nonrepeatingStitches.contains(where: { $0.name == stitch }) {
+        return .success(stitch)
+    }
+    let isRepeatingStitch = repeatingStitches.first(where: { stitch.starts(with: $0.name )})
+    if let isRepeatingStitch = isRepeatingStitch {
```
[the remark this answers is not in the document]
Comment (hnlee) [reply Sources/SwiftCliCore/StitchLibrary.swift:60]: This is Swift convention for unwrapping an optional; I agree it's a little weird but it's the language idiom.

File: Sources/SwiftCliCore/StitchLibrary.swift [hunk 2]
```diff
@@ -50,3 +48,33 @@ func isStitchValid(stitch: String) -> Bool {
     }
     return false
 }
+
+func isStitchAndStitchCountValid(stitch: String, rowNumber: Int, stitchIndex: Int) -> Result<String, InputError> {
```
Comment (hnlee) [inline Sources/SwiftCliCore/StitchLibrary.swift:52]: This method is quite long and a little hard to follow. Could you take a stab at breaking it up?

File: Tests/SwiftCliCoreTests/InputValidatorTest.swift [hunk 3]
```diff
@@ -14,7 +14,16 @@ class ValidatorTests: XCTestCase {
...
+        let err = InputError.multipleErrors
+        (errors:
+            [
+                SwiftCliCore.InputError.invalidStitch(
+                    invalidStitch: "g1",
+                    rowLocation: Optional(1),
+                    stitchIndexInRow: Optional(1)
+                )
+            ]
+        )
```
Comment (hnlee) [inline Tests/SwiftCliCoreTests/InputValidatorTest.swift:26]: I had to stare at this quite a bit, wondering what the parentheses were doing. I think you mean 

```swift
let err = InputError.multipleErrors(
    errors: [
        ...
    ]
)
```

File: Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift [hunk 4]
```diff
@@ -24,29 +24,8 @@ class MultipleStitchExpanderTest: XCTestCase {
...
-        .to(throwError(
-            InputError.invalidStitchNumber(
-                    rowNumber: nil,
-                    invalidStitch: "kx",
-                    validStitchType: "k",
-                    invalidStitchNumber: "x")
-        ))
+    func testExpandInvalidMultipleStitch() throws {
+        let result = try NestedArrayBuilder().expandRow(row: ["k0", "p1"])
+        expect(result).to(equal(["k0", "p1"]))
```
Comment (hnlee) [inline Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift:29]: A little unclear on what's going on here. Shouldn't this be throwing an error, not just returning?

**Review by hnlee — CHANGES_REQUESTED** [review]
I'm okay with you merging this in as is, but if you have time, I would like to see a follow-up PR using `[Result]`.

File: Sources/SwiftCliCore/InputValidator.swift [hunk 5]
```diff
@@ -71,14 +66,42 @@ public class InputValidator {
...
         }
+
         return .success(stitchRow)
     }
 
+    private func validateEachStitchInWholePattern(pattern: [[String]]) -> Result<[[String]], InputError> {
+        var errorArray: [InputError] = []
+        for (rowIndex, row) in pattern.enumerated() {
+            for (stitchIndex, stitch) in row.enumerated() {
+                let result = isStitchAndStitchCountValid(stitch: stitch, rowNumber: rowIndex + 1, stitchIndex: stitchIndex + 1)
```
[the remark this answers is not in the document]
Comment (graceolivia) [reply Sources/SwiftCliCore/InputValidator.swift:87]: I agree that the StitchLibrary should be a class, I will make a refactoring story for this for sure 👍

File: Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift (hunk 4 above)
Comment (graceolivia) [reply Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift:29]: This test is misleading at this point. NestedArrayBuilder used to throw errors, but it no longer does that - it simply leaves incorrect stitches to be caught after I run the next function in `InputValidator.validateInput`. Should I remove this test because it is confusing? Or is it better to throw errors within `NestedArrayBuilder` again?  I wonder what makes the most sense?

Comment (graceolivia) [reply Tests/SwiftCliCoreTests/NestedArrayBuilderTest.swift:29]: I'm going to remove this test for now because `NestedArrayBuilder` is no longer responsible for finding errors and it's confusing.

File: Tests/SwiftCliCoreTests/InputValidatorTest.swift [hunk 6]
```diff
@@ -49,8 +58,50 @@ class ValidatorTests: XCTestCase {
...
+        expect {
+            try self.inputValidator.validateInput(
+                pattern: multipleIncorrectStitchesPattern)
+        }
+        .to(throwError(error))
+    }
 
+    func testZeroCountStitchShouldThrowMultipleErrors() throws {
+        let multipleIncorrectStitchesPattern = ["k0 p1"]
+        let error = InputError.multipleErrors(errors: [
```
[the remark this answers is not in the document]
Comment (graceolivia) [reply Tests/SwiftCliCoreTests/InputValidatorTest.swift]: It's multiple errors stored in one error! A little confusing, I know...

File: Sources/SwiftCliCore/StitchLibrary.swift [hunk 7]
```diff
@@ -50,31 +50,46 @@ func isStitchValid(stitch: String) -> Bool {
...
 
-        guard repeatNumber >= 1 else {
-            return .failure(InputError.invalidStitchNumber(rowNumber: rowNumber, invalidStitch: stitch, validStitchType: stitchName, invalidStitchNumber: stitchNameSuffix, stitchIndexInRow: stitchIndex))
-        }
+}
 
-        var stitchArray: [String] = []
-        stitchArray.append(contentsOf: repeatElement((stitchName + "1"), count: repeatNumber))
-        return .success(stitch)
+private func determineIfRepeatingStitchIsACorrectRepeatingStitch(stitch: String, stitchType: StitchType, rowNumber: Int, stitchIndex: Int) -> Result<String, InputError> {
```
Comment (hnlee) [inline Sources/SwiftCliCore/StitchLibrary.swift:66]: Maybe we can just say `isValidRepeatingStitch`?

**Review by hnlee — APPROVED** [review]
Rename suggestion but feel free to merge after you make that small change!
